lib/liquid/context.rb

```diff
@@ -207,9 +202,9 @@ def lookup_and_evaluate(obj, key, raise_on_not_found: true)
       value = obj[key]
 
       if value.is_a?(Proc) && obj.respond_to?(:[]=)
-        obj[key] = value.arity == 0 ? value.call : value.call(self)
+        obj[key] = contextualize(value)
```

Comment:
I am not changing the memoization in this PR as it let to too large of a change with even more unrelated code alterations. 

I feel we should look at:
- Exploring memoization everywhere (eg.: for code path like StandardFilters to benefit from it)
- Not only memoize `if value.is_a?(Proc)`. There are some test failures appearing when trying to remove the condition, can be iterated further in subsequent work if required.